broadinstitute/single_cell_portal_core
# Service worker cache for faster local frontend development (SCP-4379)

This speeds up frontend-only development. Cycle iteration times are often 5 seconds (~25%) faster.

### Background
Previously, backend SCP API responses always hit the server.  Since these responses are equivalent when changing only frontend code, such GET requests were usually basically wasted time.  Now -- using [service worker cache](https://web.dev/learn/pwa/caching/) (aka web platform `Cache` or `CacheStorage` API) -- we can store those responses on our local web browser client, and completely eliminate backend processing time for them.

### Impact
This can save around 5 seconds per local page load, [reducing e.g. time to first scatter plot from 17.5 s to 12.6 s (28%)](https://mixpanel.com/s/3CUjc6).  It mainly optimizes frontend cycle times not covered by Vite hot module replacement (HMR).  That said, it can also benefit those if e.g. the updated component triggers downstream SCP API requests.

### Configuration
Service worker (SW) caching for local development is currently opt-in.  

The following basic Vite command won't have caching:
```
bin/vite dev
```

To enable SW caching, use this new Vite command in your local Terminal:
```
VITE_FRONTEND_SERVICE_WORKER_CACHE="true" bin/vite dev
```

You'll also notice a new file `version.txt` in our root directory.  This is used to configure the SW cache key, and as part clearing old caches.  It's used only locally right now, so it won't affect staging or production.  The value is now also logged as `scpVersion` for all development Mixpanel `app-frontend` events, which might be worth refining and extending to our other environments and loggers.

### How to clear
Clear this cache like any other service worker cache:
* Open DevTools
* Go to Application -> Cache
* Click on the latest `scp-...` key (e.g. `scp-development-1.20.0`)
* Click the "x" button to delete whole cache

You can also delete specific cache entries by highlighting and deleting them in the same view, at right.

<img width="1680" alt="Service_worker_cache_for_faster_local_frontend_development__SCP_2022-07-11" src="https://user-images.githubusercontent.com/1334561/178354694-a821f1ba-2687-4fb9-918f-7bfd5625e1a1.png">

### Analytics
Whether SW cache was enabled and hit is now recorded as part of Mixpanel events, e.g. in:
* `'perfTime:serviceWorkerCacheHit': true`
* `'perfTime:cache': 'service-worker'`

Values for `perfTime:backend` and other eliminated phases are now -1, re-using a pattern from plot data caching (#1036).

### Testing

An automated test verifies some caching here.  To manually test:

1. `yarn install` to get new JS dependencies
2. Stop Vite, then run in terminal: `VITE_FRONTEND_SERVICE_WORKER_CACHE="true" bin/vite dev`
3. Go to home page
4. Open DevTools, go to Application -> Cache
5. Confirm a cache `scp-development-1.20.0` appears and is populated with some SCP API requests
6. Go to Network panel, filter to "cell/api", confirm some SCP API requests were sent
7. Refresh home page
8. Go to Network panel, filter to "cell/api", confirm some SCP API requests were not sent
9. Go to Console panel, click "Default levels" dropdown, select "Verbose"
10. Confirm some messages in Console begin "Service worker cache hit"
11. Stop Vite, then run in terminal: `bin/vite dev`
12. Refresh home page.  If you get an error, refresh once more.
13. Go to Network panel, filter to "cell/api", confirm some SCP API requests were sent
14. Confirm no messages in Console begin "Service worker cache"
15. Deselect "Verbose" logging in Console levels dropdown (resetting to state before step 9)

This satisfies SCP-4379.
